### PR TITLE
Remove kmalloc from STM32 DMAMUX driver

### PR DESCRIPTION
**Pull request by lowlander**
This PR removes kmalloc from the stm32 dmamux driver and updates the STM32L4R5 to use the DMAMUX. 

Since I only have a nucleo_l4r5zi that is the target I do my testing with, but this should also be tested on stm32wb hardware, anybody with that hardware please give this PR a try.

### CODE REVIEW

**Review by tbursztyka — APPROVED** [review]

**Comment by FRASTM** [comment]
The HEAP_MEM_POOL_SIZE definition in the `drivers/dma/Kconfig.stm32`  could also be removed.

**Comment by FRASTM** [comment]
Could the target board be set in another PR ?

**Comment by lowlander** [comment]
> Could the target board be set in another PR ?

So just the first 2 commits in this PR and the rest in a new PR ?

**Comment by lowlander** [comment]
@FRASTM how would you like me to split this PR up? Just the first two commits in the PR, and the rest in a new one ?

**Comment by lochej** [comment]
> 
> 
> @FRASTM how would you like me to split this PR up? Just the first two commits in the PR, and the rest in a new one ?

I think that's should be that way, having DMA enabled and SPI fixed for the L4R5ZI is independant from your DMA improvement commits. Then the 2 PR could be merged separately with no issues.

File: drivers/dma/dmamux_stm32.c
Comment (lochej) [inline drivers/dma/dmamux_stm32.c:198]: I'm ok with the current name as for the channel array variable, but I would have put it in plural form: 
```
dmamux_stm32_channels_##index
```

**Comment by lowlander** [comment]
> > @FRASTM how would you like me to split this PR up? Just the first two commits in the PR, and the rest in a new one ?
> 
> I think that's should be that way, having DMA enabled and SPI fixed for the L4R5ZI is independant from your DMA improvement commits. Then the 2 PR could be merged separately with no issues.

I have no problem with splitting it up, just want to clarify why they are together, the L4R5 and WB55 seem to be the only two boards (in Zephyr) that have a CPU that has a DMAMUX. So to test this PR, that board support is needed.

**Comment by lochej** [comment]
> 
> 
> > > @FRASTM how would you like me to split this PR up? Just the first two commits in the PR, and the rest in a new one ?
> > 
> > 
> > I think that's should be that way, having DMA enabled and SPI fixed for the L4R5ZI is independant from your DMA improvement commits. Then the 2 PR could be merged separately with no issues.
> 
> I have no problem with splitting it up, just want to clarify why they are together, the L4R5 and WB55 seem to be the only two boards (in Zephyr) that have a CPU that has a DMAMUX. So to test this PR, that board support is needed.

Alright, I see what you mean, in fact the L4R5 could be merged before and tests would pass as it would just use the kmalloc approach. But then your DMAMUX update needs that first PR so the tests scenarios can be executed for the L4R5 board.

**Comment by FRASTM** [comment]
I'm fine with this approach

NB stm32g071 and stm32g4xx (nucleo_g071rb and nucleo_g4xx boards) also support DMAMUX

**Review by FRASTM — APPROVED** [review]

File: drivers/dma/dmamux_stm32.c
Comment (lowlander) [reply drivers/dma/dmamux_stm32.c:198]: As always naming things it the hardest part of programming :-)

**Comment by lowlander** [comment]
> I'm fine with this approach
> 
> NB stm32g071 and stm32g4xx (nucleo_g071rb and nucleo_g4xx boards) also support DMAMUX

Ah OK, since a few days I actually have a nucleo G474RE So could add the test to that board also.

Let me change the variable name @lochej  requested and repush this PR.

**Comment by lowlander** [comment]
@FRASTM I am doing some extra testing on my new Nucleo G474RE board, so I will need some extra time, this PR is not allowed in 2.4 anyway (because it is not a bug fix), right ?

**Comment by lowlander** [comment]
Probably would be good to wait for #28268 to be done.

**Review by tbursztyka — APPROVED** [review]

**Comment by lowlander** [comment]
@FRASTM and @tbursztyka I rerequested your reviews, because there are a lot of changes to this PR and I don't want to have "false positive" reviews. 

Also until #28268 is done this PR is a bit of a collection area for things that need to be split up in multiple PR's .

**Comment by lowlander** [comment]
@tbursztyka do you have some auto-approve script, you re-approved faster than I could type my explanation for the request :-)

**Review by FRASTM — APPROVED** [review]

**Comment by tbursztyka** [comment]
@lowlander you got me ;)
No I did look at it. The big parts is this kmalloc stuff, rest is trivial and seems legit.

**Comment by erwango** [comment]
@lowlander Is that still WIP ?

**Comment by lowlander** [comment]
> @lowlander Is that still WIP ?

@erwango Basically just waiting for #28268  to land in master so I can rebase on that.

**Comment by lowlander** [comment]
@FRASTM could you try this on a stm32wb , I tried L4R5 which seems to work.

**Comment by lowlander** [comment]
@erwango if someone can test this on a stm32wb and everything is OK I think this PR is good to go.

**Comment by FRASTM** [comment]
successfully tested for stm32wb55, on nucleo_wb55rg with PA6 = PA7 (SPI1_TX = SPI1_RX) on connector CN10 of nucleo board
```
*** Booting Zephyr OS build zephyr-v2.4.0-1463-gf4fcb165ff98  ***               
Running test suite test_spi                                                     
===================================================================             
START - test_spi_loopback                                                       
I: SPI test on buffers TX/RX 0x200001a5/0x20000a32                              
I: SPI test slow config                                                         
I: Start complete multiple                                                      
I: CS control inhibited (no GPIO device)                                        
I: Passed                                                                       
I: Start complete loop                                                          
I: Passed                                                                       
I: Start null tx                                                                
I: Passed                                                                       
I: Start half start                                                             
I: Passed                                                                       
I: Skip half end                                                                
I: Skip every 4                                                                 
I: SPI test fast config                                                         
I: Start complete multiple                                                      
I: CS control inhibited (no GPIO device)                                        
I: Passed                                                                       
I: Start complete loop                                                          
I: Passed                                                                       
I: Start null tx                                                                
I: Passed                                                                       
I: Start half start                                                             
I: Passed                                                                       
I: Skip half end                                                                
I: Skip every 4                                                                 
I: Start complete loop                                                          
I: CS control inhibited (no GPIO device)                                        
I: Passed                                                                       
I: Start complete loop                                                          
I: CS control inhibited (no GPIO device)                                        
I: Passed                                                                       
I: All tx/rx passed                                                             
 PASS - test_spi_loopback                                                       
===================================================================             
Test suite test_spi succeeded                                                   
===================================================================             
PROJECT EXECUTION SUCCESSFUL
```

**Comment by FRASTM** [comment]
however this test does not support

**Comment by lowlander** [comment]
@FRASTM 

> However, this does not support PR _drivers: dma: stm32 dmamux device is passed to the dma device_ #30041

What do you mean by that? Is that PR needed for this PR to be merged ?

**Comment by FRASTM** [comment]
This is due to a side effect of the https://github.com/zephyrproject-rtos/zephyr/pull/30041

**Comment by lowlander** [comment]
> This is due to a side effect of the #30041

@FRASTM so you want to do #30041 first? I don't have a problem with that, although I would like to merge this PR at some point, because endlessly rebasing also costs time.

But you and @erwango have to let me know what the plan is, like I said I don't want to keep rebasing this every few days to hear there is yet another PR that has to go first.

File: dts/arm/st/l4/stm32l4.dtsi
Comment (erwango) [inline dts/arm/st/l4/stm32l4.dtsi:339]: DMAMUX is not present on all L4. It's only available on the L4R, L4S and L4P versions.
So I don't think this property should be added here (even if it doesn't hurt).

File: dts/arm/st/g4/stm32g47X.dtsi
Comment (erwango) [inline dts/arm/st/g4/stm32g47X.dtsi:48]: Why not adding these directly in stm32g4.dtsi ?

File: soc/arm/st_stm32/stm32l4/Kconfig.defconfig.stm32l4r5xx
Comment (erwango) [inline soc/arm/st_stm32/stm32l4/Kconfig.defconfig.stm32l4r5xx:14]: Already selected in https://github.com/zephyrproject-rtos/zephyr/blob/47ebde30b9ea0efded1faf7920bb3a72f0d3664a/drivers/dma/Kconfig.stm32#L28

File: soc/arm/st_stm32/stm32g4/Kconfig.defconfig.stm32g431rb
Comment (erwango) [inline soc/arm/st_stm32/stm32g4/Kconfig.defconfig.stm32g431rb:14]: DMAMUX is available on whole G4 series, I think this could safely be selected in soc/arm/st_stm32/stm32g4/Kconfig.defconfig.series

File: soc/arm/st_stm32/stm32g4/Kconfig.defconfig.stm32g474re
Comment (erwango) [inline soc/arm/st_stm32/stm32g4/Kconfig.defconfig.stm32g474re:14]: ditto

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this change.
Some comments on DMAMUX selection for g4. It is supported on the whole series, so you can go safely activate it on whole series.

Btw, maybe we should make thing more simple and enable CONFIG_DMAMUX based on dts compatible detection. This would be similar to what is done on I2C for instance:
```
DT_COMPAT_ST_STM32_I2C_V1 := st,stm32-i2c-v1
config I2C_STM32_V1
	bool
	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_I2C_V1))
```
This would help getting rid of Kconfig boiler plate (that might even not be in sync with dts).

File: dts/arm/st/l4/stm32l4.dtsi
Comment (lowlander) [reply dts/arm/st/l4/stm32l4.dtsi:339]: Ah I added that becuase it didn't build on my L4R5 without it, didn't realize not all L4's need it, will take a look at where to place it so it only is present for SOC's that support it.

File: dts/arm/st/g4/stm32g47X.dtsi
Comment (lowlander) [reply dts/arm/st/g4/stm32g47X.dtsi:48]: I think I had a reason for that at the time, I will have to look into it again to see if it can be put in the general G4 file.

**Comment by erwango** [comment]
@lowlander btw, this PR has shouldn't have dependency with #30041, so you can keep going on.

File: soc/arm/st_stm32/stm32l4/Kconfig.defconfig.stm32l4r5xx
Comment (FRASTM) [inline soc/arm/st_stm32/stm32l4/Kconfig.defconfig.stm32l4r5xx:17]: This will conflict with the #30784
As the `drivers/dma/Kconfig.stm32` is configuring the DMAMUX_STM32 for the series

File: dts/arm/st/g4/stm32g47X.dtsi
Comment (lowlander) [reply dts/arm/st/g4/stm32g47X.dtsi:48]: @erwango they both have different values for the number of channels 12 vs 16.

File: dts/arm/st/l4/stm32l4.dtsi
Comment (erwango) [reply dts/arm/st/l4/stm32l4.dtsi:339]: I tend to prefer this: https://github.com/zephyrproject-rtos/zephyr/pull/30784/files#diff-cc2c619a82c50138c0684c40e7fa9e96a9416475c575dbf7ee64709fa7798a23.

**Review by erwango — COMMENTED** [review]
@lowlander txs for the changes.

One comment on dts part, for which I prefer the way it is handled in overlapping PR: https://github.com/zephyrproject-rtos/zephyr/pull/30784

More generally, maybe you could rebase on top of https://github.com/zephyrproject-rtos/zephyr/pull/30784 ?

Then, fyi, I'm waiting these PRs to be merged before switching to Kconfig symbol selection txs to dts for this driver (to simplify the Kconfig hell).

File: dts/arm/st/l4/stm32l4r5.dtsi
Comment (shlomow) [inline dts/arm/st/l4/stm32l4r5.dtsi:247]: should be 89

**Review by shlomow — CHANGES_REQUESTED** [review]

File: dts/arm/st/l4/stm32l4r5.dtsi
Comment (lowlander) [reply dts/arm/st/l4/stm32l4r5.dtsi:247]: @shlomow what should be 89 ?

Comment (shlomow) [reply dts/arm/st/l4/stm32l4r5.dtsi:247]: @lowlander the `dma-requests`

Comment (FRASTM) [reply dts/arm/st/l4/stm32l4r5.dtsi:247]: + space mark ;-)  `dma-requests = <89>;`

File: dts/arm/st/g4/stm32g431.dtsi
Comment (erwango) [inline dts/arm/st/g4/stm32g431.dtsi:7]: I'm not fond of adding this intermediary files (stm32g43X and stm32g47X).
Also, I'd prefer we populate g4.dtsi with one version of the nodes and we overwrite the diverging property when required (say in stm32g474.dtsi), similarly to https://github.com/zephyrproject-rtos/zephyr/pull/30784.

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/l4/stm32l4r5.dtsi
Comment (FRASTM) [reply dts/arm/st/l4/stm32l4r5.dtsi:247]: `<dma-requests>` is the Number of DMAMUX peripheral request inputs 
= 89 (STM32L4Rxxx and STM32L4Sxxx devices)
= 90 (STM32L4P5xx and STM32L4Q5xx devices)

(_26_ is the Number of DMAMUX request trigger inputs which is not relevant for this driver)

Comment (lowlander) [reply dts/arm/st/l4/stm32l4r5.dtsi:247]: This part will be "gone" from this PR after #30784 has been merged.

File: soc/arm/st_stm32/stm32g4/Kconfig.defconfig.series
Comment (lowlander) [inline soc/arm/st_stm32/stm32g4/Kconfig.defconfig.series:16]: With the new kconfig thing is this still needed? @FRASTM  @erwango ?

Comment (FRASTM) [reply soc/arm/st_stm32/stm32g4/Kconfig.defconfig.series:16]: no more, because of  the #31163

Comment (lowlander) [reply soc/arm/st_stm32/stm32g4/Kconfig.defconfig.series]: @FRASTM Well something isn't working with out it, any ideas ?

Comment (shlomow) [reply soc/arm/st_stm32/stm32g4/Kconfig.defconfig.series]: Instead of `compatible = "st,stm32-dma"` it should be `compatible = "st,stm32-dma-v2"`

Comment (lowlander) [reply soc/arm/st_stm32/stm32g4/Kconfig.defconfig.series]: @shlomow thanx, fixed

File: tests/drivers/spi/spi_loopback/boards/nucleo_l4r5zi.conf
Comment (erwango) [inline tests/drivers/spi/spi_loopback/boards/nucleo_l4r5zi.conf:3]: Not needed, selected by `SPI_STM32_DMA`

File: tests/drivers/spi/spi_loopback/boards/nucleo_g474re.conf
Comment (erwango) [inline tests/drivers/spi/spi_loopback/boards/nucleo_g474re.conf:3]: Ditto

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/g4/stm32g47X.dtsi
Comment (erwango) [inline dts/arm/st/g4/stm32g47X.dtsi:7]: Last blocking point:
I still consider this file addition is not requested.
If required, as done on L4 series, you can overwrite dma1 and dma2 property easily without any file addition.

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/g4/stm32g47X.dtsi
Comment (erwango) [reply dts/arm/st/g4/stm32g47X.dtsi:7]: @lowlander if these diverge too much, let's put dma1/2 nodes definitions in stm32g431.dtsi and stm32g471.dtsi.
But I don't see a need for new intermediate files

File: dts/arm/st/g4/stm32g431.dtsi
Comment (erwango) [inline dts/arm/st/g4/stm32g431.dtsi:2]: Not sure this change is valid.

**Comment by lowlander** [comment]
@erwango @nashif the no ; after macros warning/error causes serious troubles. Without the ; compilation fails deep inside some other macro expansions, in code I never touched (the ; also has been there way before my changes BTW)

File: dts/arm/st/g4/stm32g47X.dtsi
Comment (lowlander) [reply dts/arm/st/g4/stm32g47X.dtsi]: @erwango is the dts OK like it is now ?

File: tests/drivers/spi/spi_loopback/boards/nucleo_l4r5zi.overlay
Comment (erwango) [inline tests/drivers/spi/spi_loopback/boards/nucleo_l4r5zi.overlay:2]: Please use your copyright

File: tests/drivers/spi/spi_loopback/boards/nucleo_g474re.overlay
Comment (erwango) [inline tests/drivers/spi/spi_loopback/boards/nucleo_g474re.overlay:2]: Please use your copyright

**Review by erwango — CHANGES_REQUESTED** [review]
Minor copyright stuff, otherwise LGTM

File: tests/drivers/spi/spi_loopback/boards/nucleo_g474re.overlay
Comment (lowlander) [reply tests/drivers/spi/spi_loopback/boards/nucleo_g474re.overlay:2]: Those where copied files, from other examples, so not really my copyright. (and to be honest I don't care to much about my copyright for FOSS projects anyway)

File: dts/arm/st/g4/stm32g431.dtsi
Comment (lowlander) [reply dts/arm/st/g4/stm32g431.dtsi:2]: Also a copied file, the "original" empty file is pretty much gone.  I never really know what to do in causes like that.

File: tests/drivers/spi/spi_loopback/boards/nucleo_g474re.overlay
Comment (erwango) [reply tests/drivers/spi/spi_loopback/boards/nucleo_g474re.overlay:2]: I get your point but my reasoning is the following:
Even in case of copy/paste I consider that there is an effort of doing the effort and testing.
Besides the version of the files you copy might be different of the original version of the file that was initially "copyrighted" 
Doing otherwise, we'd end up having copyright info that totally offset vs reality, and these could be used for various stats, which are valuable to the project, when they are reflect reality.

But ok, I won't block for this.

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
Approved, but please fix CI

File: tests/drivers/spi/spi_loopback/boards/nucleo_g474re.overlay
Comment (lowlander) [reply tests/drivers/spi/spi_loopback/boards/nucleo_g474re.overlay:2]: I have no problem changing it, but I just feel weird when I copy a file and remove 1 or 2 lines and than put my copyright in it.  On the other hand that is mostly with very trivial files I doubt the original copyright holder would have a problem with it, in the end the license is what is most important.  

But is there any "global" Zephyr guidance for cases like this ? 

Let me put back the "Richard Osterloh <richard.osterloh@gmail.com>" copyright because that file was already there and put my copyright in the new files (even if they are copied, but I changed them)

**Comment by lowlander** [comment]
> Approved, but please fix CI

@erwango see the msg above, I doubt I can fix the CI "; after macro" problem because without the ; the build fails deep in Zephyr init macros, in code that I never touched. Hence the ping to  @nashif  for some guidance on this CI error.

File: drivers/dma/dmamux_stm32.c
Comment (lowlander) [inline drivers/dma/dmamux_stm32.c:229]: @erwango @nashif CI complains about this ; on line 228, but without it compilation fails inside the `DEVICE_DT_INST_DEFINE` macro. The ; is used in about all files that also use `DEVICE_DT_INST_DEFINE` and I don't know how to fix it.

Comment (erwango) [reply drivers/dma/dmamux_stm32.c:229]: No clue. @galak maybe ?

Comment (str4t0m) [reply drivers/dma/dmamux_stm32.c:229]: #31261 seems to have the same problem.

**Pull request merged**
